core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArrayAggregationAfterDistinct.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return project()
+                .matching(entry -> hasArrayDistinct(metadata, entry))
+                .with(source().matching(aggregation()
+                        .matching(ArrayAggregationAfterDistinct::hasArrayAggregation)
+                        .capturedAs(CHILD)));
+    }
+
+    private static boolean hasArrayAggregation(AggregationNode aggregationNode)
+    {
+        return aggregationNode.getAggregations().values().stream()
+                .anyMatch(aggregation -> aggregation.getResolvedFunction().getSignature().getName().equals(ArrayAggregationFunction.NAME) && aggregation.getArguments().size() == 1);
+    }
+
+    private static boolean hasArrayDistinct(Metadata metadata, ProjectNode projectNode)
+    {
+        return projectNode.getAssignments().getExpressions().stream()
+                .anyMatch(assignment -> assignment instanceof FunctionCall &&
+                        metadata.decodeFunction(((FunctionCall) assignment).getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME));
+    }
+
+    @Override
+    public Result apply(ProjectNode projectNode, Captures captures, Context context)
+    {
+        AggregationNode aggregationNode = captures.get(CHILD);
+
+        Assignments.Builder assignmentsBuilder = Assignments.builder();
+        ImmutableMap.Builder<Symbol, Aggregation> aggregationsBuilder = ImmutableMap.builder();
+
+        List<Symbol> matchedAggregations = Lists.newArrayList();
+        List<Symbol> matchedAssignments = Lists.newArrayList();
+
+        for (Map.Entry<Symbol, Expression> entry : projectNode.getAssignments().entrySet()) {
+            if (entry.getValue() instanceof FunctionCall) {
+                Symbol symbol = entry.getKey();
+                FunctionCall functionCall = (FunctionCall) entry.getValue();
+                ResolvedFunction resolvedFunction = metadata.decodeFunction(functionCall.getName());
+                if (resolvedFunction.getSignature().getName().equals(ArrayDistinctFunction.NAME)) {
+                    SymbolReference argument = (SymbolReference) (getOnlyElement(functionCall.getArguments()));
+                    Optional<Map.Entry<Symbol, Aggregation>> matchingSymbolAndAggregation = aggregationNode.getAggregations().entrySet().stream()
+                            .filter(entry1 -> entry1.getKey().getName().equals(argument.getName())).findFirst();
+                    if (matchingSymbolAndAggregation.isPresent()) {
+                        Map.Entry<Symbol, Aggregation> symbolAggregationEntry = matchingSymbolAndAggregation.get();
+                        Aggregation aggregation = symbolAggregationEntry.getValue();
+                        Symbol aggregationSymbol = symbolAggregationEntry.getKey();
+                        if (!aggregation.isDistinct()) {
+                            // recreate aggregation with distinct
```

Comment:
Please note that one aggregation could be referenced in multiple ProjectNode's assignments. If you change it into `distinct`, you break other assignments.